File: yht_custom/sales_flow.py
# Copyright (c) 2026, Enfono Technologies and contributors
# For license information, please see license.txt

"""Sales and purchase flow policy.

MoM §2.2 and the client's own module sheet settle three things that the legacy
site left to each user's discretion:

* **Stock is not moved twice.** `update_stock` is switched off when — and only
  when — the goods on the document have already moved on a stock document.

  🔴 THIS REVERSES "Delivery Note is compulsory" (MoM §2.2), which forced
  `update_stock = 0` for every non-bypass user until 2026-09-10. Client sheet
  item 25 asks for a standalone invoice to open TICKED for branch users too, and
  the gate approved it (Q1) on the second asking. What is given up, stated: on
  the legacy site 1,339 of 2,337 submitted invoices came through a Delivery Note
  and 998 updated stock directly, and the DN-compulsory rule existed to end that
  split. A branch user can again bill and ship in one document. The *scope*
  boundary is untouched — `branch_guard.validate_branch_scope` still rejects a
  warehouse outside the branch — what changed is the flow policy.
* **A zero-rate Delivery Note is blocked.** Goods must not leave at nil value.
* **A submitted Delivery Note is locked**, so the printed copy the driver carries
  always matches the record.
* **Purchase stock arrives on the Purchase Receipt**, not on the Purchase Invoice.

The measured reason this matters: of 2,337 submitted invoices on the legacy site,
1,339 came through a Delivery Note and 998 updated stock directly. Same business,
two incompatible habits, and no way to reconcile a delivery against an invoice.
"""

import frappe
from frappe import _
from frappe.model import display_fieldtypes, table_fields
from frappe.utils import cint, cstr, flt

from yht_custom import branch_defaults, other_remarks

#: Roles that may still tick `update_stock` on a Sales Invoice, i.e. bill and
#: ship in one document. Everyone else goes through a Delivery Note.
#:
#: ⚠️ THESE NO LONGER DECIDE `update_stock` (client sheet item 25, gate Q1).
#: `lock_submitted_delivery_note` and the Delivery Note refresh comment still
#: read them, which is why they are kept.
DIRECT_STOCK_ROLES = (
	"System Manager",
	"Stock Manager",
	"Sales Manager",
	"Sales Master Manager",
	"Accounts Manager",
)

#: Ceiling on `sales_order_has_stock_document`'s caller-supplied row list.
#:
#: The endpoint answers an existence question about row names it does not
#: permission-check individually, so the list it will accept is bounded to keep
#: the leak at "one bit per guessed name" rather than "walk the table".
#: Measured on `yht-test` 2026-09-10: the largest document in the site's history
#: is 169 Sales Invoice rows, 70 Sales Order rows, 54 Purchase Invoice rows. 500
#: leaves generous headroom over all three and still refuses a scripted probe.
MAX_PROBED_ROWS = 500

#: Client sheet item 25 — how each invoice reaches the stock document behind it.
#:
#: `row_link` is the child field ERPNext writes when the goods have ALREADY moved
#: on a stock document; `order_link` is the row's link back to the order, and
#: `stock_order_link` is that same order row seen from the stock document's side.
#: The two sides are NOT symmetric — `Delivery Note Item.so_detail` against
#: `Purchase Receipt Item.purchase_order_item` — so both are named rather than
#: derived, and every one was read off live meta before it was written here.
_STOCK_ROUTE = {
	"Sales Invoice": {
		"row_link": "dn_detail",
		"order_link": "so_detail",
		"stock_parent": "Delivery Note",
		"stock_child": "Delivery Note Item",
		"stock_order_link": "so_detail",
	},
	"Purchase Invoice": {
		"row_link": "pr_detail",
		"order_link": "po_detail",
		"stock_parent": "Purchase Receipt",
		"stock_child": "Purchase Receipt Item",
		"stock_order_link": "purchase_order_item",
	},
}


def _may_bypass(user=None) -> bool:
	user = user or frappe.session.user
	if user == "Administrator":
		return True
	return bool(set(frappe.get_roles(user)) & set(DIRECT_STOCK_ROLES))


def _stock_document_against(doctype: str, order_rows) -> str | None:
	"""The stock document already covering any of these order rows, or None.

	ONE query however many rows the document carries. An N+1 here would run on
	every save of every invoice, and a twenty-line invoice is ordinary.
	"""
	spec = _STOCK_ROUTE.get(doctype)
	if not spec or not order_rows:
		return None

	child = frappe.qb.DocType(spec["stock_child"])
	parent = frappe.qb.DocType(spec["stock_parent"])
	found = (
		frappe.qb.from_(child)
		.join(parent)
		.on(parent.name == child.parent)
		.select(parent.name)
		.where(
			# 🔴 SUBSCRIPT, NEVER `.field(...)`. `frappe/query_builder/__init__.py:23`
			# monkey-patches pypika with `Selectable.field = PseudoColumn("field")` — an
			# INSTANCE, not the method pypika ships. `child.field` therefore never reaches
			# the `__getattr__` on line 21 (it is a real class attribute) and calling it
			# raises `TypeError: 'PseudoColumn' object is not callable`. Line 22 patches
			# `__getitem__` to build the `Field` directly, which is why this form works and
			# is the only form used anywhere in frappe, erpnext or this app.
			child[spec["stock_order_link"]].isin(list(order_rows))
			& (parent.docstatus == 1)
			& (parent.is_return == 0)
		)
		.limit(1)
	).run()
	return found[0][0] if found else None


def _stock_already_moved(doc) -> str | None:
	"""Item 25's linkage rule. Zeroes `update_stock`; names the document if visible.

	🔴 THIS REPLACED A ROLE TEST, AND THAT IS THE WHOLE ITEM (gate Q1). Until now
	`update_stock` was forced to 0 for every non-bypass user, which is the MoM
	§2.2 "Delivery Note is compulsory" policy. The client has asked twice for a
	standalone invoice to open TICKED for branch users too, so the question the
	rule asks is no longer *who is keying this in* but *have these goods already
	moved*:

	1. a row carrying the stock-document link → 0, and return None, because
	   ERPNext's own `depends_on` HIDES the box in that state: there is nothing
	   for the operator to reconcile and a message would only confuse.
	2. otherwise, an order row behind this document that a submitted, non-return
	   stock document already covers → 0, and return that document's name so the
	   caller can say so. Here the field IS visible and the flip needs explaining.
	3. otherwise leave the value alone, the new default included.

	Step 1 is not cosmetic. `get_mapped_doc` copies field defaults onto the
	target, so without it every invoice mapped from a Delivery Note would arrive
	with `update_stock = 1` and `SalesInvoice.validate_delivery_note()` would
	refuse the save with an error the operator cannot act on — the field being
	hidden. The purchase side has no equivalent throw at all, so the stock would
	simply be received a second time, silently.
	"""
	spec = _STOCK_ROUTE[doc.doctype]
	rows = doc.get("items") or []

	if any(cstr(row.get(spec["row_link"])).strip() for row in rows):
		doc.update_stock = 0
		return None

	order_rows = {
		cstr(row.get(spec["order_link"])).strip()
		for row in rows
		if cstr(row.get(spec["order_link"])).strip()
	}
	stock_document = _stock_document_against(doc.doctype, order_rows)
	if not stock_document:
		return None

	doc.update_stock = 0
	return stock_document


# ------------------------------------------------------------- Sales Invoice


def enforce_delivery_note_route(doc, method=None):
	"""Switch `update_stock` off when the goods have already gone out on a note.

	Server-side rather than JS-only: the form lock stops the ordinary user, this
	stops a REST call, an import and a Server Script as well. See
	`_stock_already_moved` for the rule and for what it replaced.

	🔴 A RETURN IS NOT A SALE, and this had no `is_return` branch until 2026-08-26.
	How the goods LEFT decides how they come back, so `update_stock` on a return is
	dictated by the original document, not by this rule — ERPNext itself throws
	when a return ticks it and its original did not. Zeroing it unconditionally
	meant a branch user returning a legacy direct-stock invoice could never bring
	the goods back at all (measured: `is_return=1, update_stock=1` in, `0` out).
	`return_flow.enforce_return_stock_route` owns the return case.
	"""
	if cint(doc.get("is_return")):
		return
	if not cint(doc.get("update_stock")):
		return

	note = _stock_already_moved(doc)
	if not note:
		return

	frappe.msgprint(
		_(
			"These goods are already going out on Delivery Note {0}, so <b>Update Stock</b> has been switched off on this invoice."
		).format(note),
		title=_("Already Delivered"),
		indicator="orange",
	)


# -------------------------------------------------------------- Delivery Note


def validate_delivery_note(doc, method=None):
	"""Block a nil-value delivery, and keep a submitted DN immutable."""
	_block_zero_rate(doc)


def _block_zero_rate(doc):
	"""Goods must not leave the warehouse at zero value.

	A free-of-charge delivery is a real thing, but it belongs on a document that
	says so — not as a silent zero on an ordinary DN, which is how stock walks out
	unaccounted.
	"""
	if cint(doc.get("is_return")):
		return  # a return legitimately mirrors the original, including nil lines

	offenders = [row.idx for row in (doc.get("items") or []) if flt(row.get("rate")) <= 0]
	if not offenders:
		return

	frappe.throw(
		_("Rows {0} have a zero rate. A Delivery Note cannot release goods at nil value.").format(
			", ".join(str(i) for i in offenders)
		),
		title=_("Zero Rate Not Allowed"),
	)


def _comparable(doc, field):
	"""One field's value, in a shape two copies of the same document can compare.

	🔴 A CHILD TABLE COMPARED DIRECTLY IS ALWAYS "CHANGED", AND THAT MADE THE LOCK
	ABSOLUTE. `doc.get("items")` and `before.get("items")` are two different lists
	holding two different child `Document` objects, and frappe's `BaseDocument`
	defines no `__eq__` — so `!=` falls back to identity and EVERY Table field
	reports a change on every save. `lock_submitted_delivery_note` therefore threw
	for any post-submit edit by a non-bypass user whatever actually changed
	(measured: `Changed: items, sales_team, taxes` on a save that touched only
	`custom_other_remarks`), which is what made item 28's exemption inert on the
	one doctype the client most wants the field on.

	A per-row dict compares by VALUE. `no_default_fields` drops `modified` /
	`modified_by` / `idx` churn that is not an operator edit, and
	`no_private_properties` drops `__unsaved` / `__islocal`, which are set on the
	in-memory copy and absent from the one loaded out of the database — on their
	own enough to make every row differ again.
	"""
	if field.fieldtype in table_fields:
		return [
			row.as_dict(no_default_fields=True, no_private_properties=True)
			for row in (doc.get(field.fieldname) or [])
		]
	return doc.get(field.fieldname)


def lock_submitted_delivery_note(doc, method=None):
	"""Refuse post-submit edits to a Delivery Note.

	ERPNext allows `allow_on_submit` fields to change after submit. On a document
	the driver has already carried out of the yard, that means the printed copy
	and the record can silently diverge.
	"""
	if doc.docstatus != 1:
		return
	before = doc.get_doc_before_save()
	if not before:
		return

	# Fields ERPNext or our own automation legitimately writes after submit.
	#
	# 🔴 `other_remarks.FIELDNAME` IS THE EXEMPTION THIS LOCK EXISTS TO GRANT.
	# Client sheet item 28 makes "Other Remarks" `allow_on_submit` precisely so a
	# note can be added to a document that has already gone out. Without it here
	# the lock refuses that edit for every non-bypass user — which is exactly who
	# the field is for, and Delivery Note is the doctype the client most wants it
	# on. Read from the module rather than retyped: a second spelling of a
	# fieldname is how an exemption quietly stops matching.
	permitted = {
		"status", "per_billed", "per_returned", "billing_status",
		"modified", "modified_by", "_user_tags", "_comments", "_assign", "_liked_by",
		"docstatus", "workflow_state",
		other_remarks.FIELDNAME,
	}
	changed = {
		field.fieldname
		for field in doc.meta.fields
		if field.fieldtype not in display_fieldtypes
		and field.fieldname not in permitted
		and _comparable(doc, field) != _comparable(before, field)
	}
	if not changed:
		return
	if _may_bypass():
		return

	frappe.throw(
		_("A submitted Delivery Note cannot be edited. Changed: {0}").format(", ".join(sorted(changed))),
		title=_("Delivery Note Locked"),
	)


# ------------------------------------------------------------ Purchase Invoice


def enforce_purchase_receipt_route(doc, method=None):
	"""Switch `update_stock` off when the goods have already been received.

	Expense invoices are exempt: they carry no stock at all, and
	`expense_invoice.before_validate` zeroes the flag for them in any case.

	A buyer's wording, not a seller's: this handler and its Sales Invoice twin
	share a rule but not a sentence — an instruction naming a Delivery Note is
	one a purchasing clerk cannot follow.

	🔴 A RETURN IS NOT A SALE, and this had no `is_return` branch until 2026-08-26.
	How the goods LEFT decides how they come back, so `update_stock` on a return is
	dictated by the original document, not by this rule — ERPNext itself throws
	when a return ticks it and its original did not. Zeroing it unconditionally
	meant a branch user returning a legacy direct-stock bill could never bring
	the goods back at all (measured: `is_return=1, update_stock=1` in, `0` out).
	`return_flow.enforce_return_stock_route` owns the return case.
	"""
	if cint(doc.get("is_return")):
		return
	if cint(doc.get("custom_is_expense_invoice")):
		return
	if not cint(doc.get("update_stock")):
		return

	receipt = _stock_already_moved(doc)
	if not receipt:
		return

	frappe.msgprint(
		_(
			"These goods are already arriving on Purchase Receipt {0}, so <b>Update Stock</b> has been switched off on this bill."
		).format(receipt),
		title=_("Already Received"),
		indicator="orange",
	)


# ------------------------------------------------------------------ whitelisted


@frappe.whitelist()
def may_use_direct_stock() -> bool:
	"""Whether the current user holds a bypass role.

	No longer decides `update_stock` (item 25) — it still drives the Delivery
	Note refresh comment, and `lock_submitted_delivery_note` reads the same
	answer.
	"""
	return _may_bypass()


@frappe.whitelist()
def sales_order_has_stock_document(doctype: str = "Sales Invoice", details=None) -> bool:
	"""Have any of these order rows already moved on a stock document?

	The second half of item 25's form lock: a row's own `dn_detail` / `pr_detail`
	is visible to the client, but "does the Sales Order behind this line already
	have a Delivery Note" is a question only the server can answer.

	A public HTTP endpoint, so be exact about what is and is not checked.

	CHECKED: read permission on `doctype` — the INVOICE doctype the form is on,
	which is what gates reaching this endpoint at all.

	NOT CHECKED: the caller-supplied `details`. Those are Sales Order Item row
	names, probed against `Delivery Note Item` / `Purchase Receipt Item` with no
	branch filter and no `permission_query_conditions` — a report bypasses those
	too (gotcha 20), and a row-level check here would be a second query per row
	on every form refresh. What that costs is bounded deliberately: the answer is
	a single boolean, never a document name, and `details` is capped at
	MAX_PROBED_ROWS so the endpoint cannot be walked as a bulk oracle. The
	exposure is therefore one bit per correctly guessed row name — a row name is
	a `hash` autoname, not a series — for a caller who already holds read on the
	invoice doctype.
	"""
	doctype = cstr(doctype)
	frappe.has_permission(doctype, "read", throw=True)

	if isinstance(details, str):
		details = frappe.parse_json(details)
	order_rows = {cstr(row).strip() for row in (details or []) if cstr(row).strip()}
	if len(order_rows) > MAX_PROBED_ROWS:
		# A real form sends one name per line. Anything past that is not a form.
		frappe.throw(
			_("Too many rows to check at once ({0}).").format(len(order_rows)),
			frappe.ValidationError,
		)

	return bool(_stock_document_against(doctype, order_rows))


# --------------------------------------------------- Get Items From > Sales Order
#
# 🔴 UPSTREAM SIGNATURE CLASH. `frappe.model.mapper.map_docs` — what the desk's
# "Get Items From" dialog posts to — passes the dialog's `args` POSITIONALLY into
# slot three:
#
#     _args = (src, target_doc, json.loads(args)) if args else (src, target_doc)
#
# and the two ERPNext mappers disagree about what slot three is (v15.119.2):
#
#     delivery_note.make_sales_invoice(source_name, target_doc=None, args=None)
#     sales_order.make_sales_invoice(source_name, target_doc=None,
#                                    ignore_permissions=False, args=None)
#
# So Delivery Note works and Sales Order does not: the dialog's payload lands on
# `ignore_permissions`, frappe's pydantic argument validation rejects a dict for
# `Union[int, bool, float]`, and the request 417s. The desk swallows it — the
# dialog closes and the invoice is simply left with no items, which is what
# "Get items from SO to Invoice not working" looks like from the outside.
#
# `map_docs` resolves `frappe.override_whitelisted_method` BEFORE calling, so a
# hook override is honoured. This shim decides by TYPE rather than by position, so
# it stays correct for both callers: the desk (a dict/JSON object in slot three)
# and any code that passes `ignore_permissions` positionally, as ERPNext's own
# signature invites.


@frappe.whitelist()
def make_sales_invoice_from_sales_order(source_name, target_doc=None, third=None, fourth=None):
	"""ERPNext's Sales Order -> Sales Invoice mapper, callable the way the desk calls it."""
	from erpnext.selling.doctype.sales_order.sales_order import (
		make_sales_invoice as erpnext_make_sales_invoice,
	)

	def _looks_like_args(value):
		if isinstance(value, dict):
			return True
		# `map_docs` json-decodes before calling, but a direct HTTP caller may not.
		return isinstance(value, str) and value.strip().startswith("{")

	if _looks_like_args(third):
		args, ignore_permissions = third, fourth
	else:
		ignore_permissions, args = third, fourth

	return erpnext_make_sales_invoice(
		source_name,
		target_doc,
		ignore_permissions=cint(ignore_permissions),
		args=args,
	)


@frappe.whitelist()
def return_naming_series(doctype: str = "Sales Invoice") -> str | None:
	"""The series a RETURN of ``doctype`` will take, for the calling user's branch.

	``None`` means "leave the picker alone": the branch configures no return series
	for this doctype, or the series could not be resolved at all — several branches
	exist and this caller is pinned to none of them. A bypass role is NOT on its own
	enough to return ``None`` any more: on a single-branch site the series falls back
	to that branch for everybody, because a credit note must not take the invoice
	counter whoever keys it in. See ``branch_defaults._branch_series_rows``.

	The entry point calls this so the form shows the series it will actually get.
	Without it the picker keeps the form's pre-filled ``KSIN-`` after ``is_return`` is
	ticked and the document saves as ``KSSR-``, which reads as a bug.
	"""
	doctype = cstr(doctype)
	frappe.has_permission(doctype, "create", throw=True)
	return branch_defaults.configured_series(doctype, is_return=1)

File: yht_custom/public/js/sales_flow.js
// Copyright (c) 2026, Enfono Technologies and contributors
// For license information, please see license.txt
//
// Delivery-Note-compulsory flow, form side.
//
// The lock here is convenience. yht_custom.sales_flow enforces the same rule on
// before_validate, so a REST call or an import cannot route around it.

frappe.provide("yht_custom.flow");

yht_custom.flow.may_direct = null;

yht_custom.flow.check = function () {
	if (yht_custom.flow.may_direct !== null) return Promise.resolve(yht_custom.flow.may_direct);
	return frappe.call({ method: "yht_custom.sales_flow.may_use_direct_stock" }).then((r) => {
		yht_custom.flow.may_direct = !!r.message;
		return yht_custom.flow.may_direct;
	});
};

// ------------------------------------------------- Update Stock, form side
//
// Client sheet item 25. This REPLACED a role-driven handler that un-ticked
// `update_stock` on every new invoice and greyed the box out for anyone without
// a bypass role. The rule is now a LINKAGE test, not a role test: a standalone
// invoice opens ticked for everybody, and the box is only taken away when the
// goods have already moved. `yht_custom.sales_flow.enforce_delivery_note_route`
// and its purchase twin decide the same thing on before_validate, so a REST
// call cannot route around this — the form's job is to say WHY, before a save.
//
// ⚠️ READ-ONLY, NEVER HIDDEN. The operator has to SEE that the box is off and
// read the reason; a field that has vanished just looks like a missing feature.
// (erpnext's own `depends_on` does hide it once a row carries the link — that
// is upstream's call, and in that state there is nothing to reconcile anyway.)
//
// `frm.set_df_property` on a HEADER field is safe. The recorded no-op is about
// GRID cells, which need the six-argument per-row form.
const STOCK_ROUTE = {
	"Sales Invoice": {
		row_link: "dn_detail",
		order_link: "so_detail",
		moved: __("These goods have already gone out on a Delivery Note."),
		ordered: __("The Sales Order behind this invoice has already been delivered."),
	},
	"Purchase Invoice": {
		row_link: "pr_detail",
		order_link: "po_detail",
		moved: __("These goods have already arrived on a Purchase Receipt."),
		ordered: __("The Purchase Order behind this bill has already been received."),
	},
};

function lock_update_stock(frm, reason) {
	frm.set_df_property("update_stock", "read_only", reason ? 1 : 0);
	frm.set_df_property("update_stock", "description", reason || "");
}

function apply_stock_route(frm) {
	const spec = STOCK_ROUTE[frm.doc.doctype];
	if (!spec) return;

	const rows = frm.doc.items || [];

	// 🔴 CLEAR THE CACHE KEY ON EVERY PATH THAT DOES NOT SET IT. It is only ever
	// written on the server-probe path below, so without this a form that HAD
	// order-linked rows keeps the old signature after those rows are removed —
	// and the next call matching that stale signature returns early at the
	// `=== signature` check, leaving the "already gone out" read-only and its
	// description on screen for a document that no longer has anything behind it.
	if (rows.some((row) => row[spec.row_link])) {
		frm.__yht_stock_route = null;
		lock_update_stock(frm, spec.moved);
		return;
	}

	const details = rows.map((row) => row[spec.order_link]).filter(Boolean);
	if (!details.length) {
		frm.__yht_stock_route = null;
		lock_update_stock(frm, null);
		return;
	}

	// Only the server can answer "has the order behind this line already
	// shipped". Cached per set of rows so a refresh storm does not become a
	// request storm.
	const signature = details.slice().sort().join("|");
	if (frm.__yht_stock_route === signature) return;
	frm.__yht_stock_route = signature;

	frappe
		.call({
			method: "yht_custom.sales_flow.sales_order_has_stock_document",
			args: { doctype: frm.doc.doctype, details: JSON.stringify(details) },
		})
		.then((r) => {
			if (frm.__yht_stock_route !== signature) return;
			lock_update_stock(frm, r && r.message ? spec.ordered : null);
		});
}

for (const doctype of Object.keys(STOCK_ROUTE)) {
	frappe.ui.form.on(doctype, {
		// Both events, and both are needed: `refresh` alone misses the first
		// paint of a mapped document, `onload` alone misses a re-render after the
		// items grid changes.
		onload(frm) {
			apply_stock_route(frm);
		},
		refresh(frm) {
			apply_stock_route(frm);
		},
	});
}

frappe.ui.form.on("Delivery Note", {
	refresh(frm) {
		if (frm.doc.docstatus !== 1) return;
		yht_custom.flow.check().then((may_direct) => {
			if (may_direct) return;
			// Submitted DNs are locked server-side; say so rather than letting the
			// user discover it by having a save rejected.
			frm.dashboard.add_comment(
				__("This Delivery Note is submitted and locked. Raise a return to correct it."),
				"orange",
				true
			);
		});
	},
});


// ------------------------------------------------------- new sales return
//
// 🔴 A SHORTCUT CANNOT TICK `is_return`, AND NEITHER CAN A URL.
//
// `Sales Invoice.is_return` is `no_copy = 1`, and `create_new.js` skips every no_copy
// field when it applies `frappe.route_options`:
//
//     if (df && !df.no_copy) doc[fieldname] = value;
//
// Measured on the site, all three obvious routes come back with is_return = 0:
//   · /app/sales-invoice/new?is_return=1   — frappe also STRIPS the query string, so a
//     client script cannot recover it either
//   · the same with a #hash                — stripped as well
//   · frappe.new_doc("Sales Invoice", { is_return: 1 })
//
// Only setting it AFTER the form exists works. Hence this helper: everything that
// offers a "Sales Return" entry point routes through it.
//
// Ticking it is what earns the separate entry point — `branch_defaults` picks the
// naming series from `is_return` at before_insert, so a BRANCH USER's document numbers
// KSSR- (credit note) instead of KSIN-, which is the whole reason an accountant wants
// the two apart. A user holding a bypass role (System Manager, Accounts Manager, Sales
// Manager, …) is exempt from that override by design and keeps whatever the picker
// shows, so the alert below does not promise a series.

frappe.provide("yht_custom.sales");

// `frappe.new_doc` resolves before the form has finished rendering, and how long that
// takes depends on where the click came from: from a list view the form is ready almost
// at once, from a Page (the branch dashboard) it is not. A single readiness check passed
// on the list and bailed out silently on the dashboard, leaving an ORDINARY invoice open
// — the operator's next click was a sale, not a credit note. Wait for the form instead.
function wait_for_new_form(doctype, tries = 40) {
	return new Promise((resolve) => {
		const tick = () => {
			const frm = window.cur_frm;
			if (frm && frm.doc && frm.doc.doctype === doctype && frm.doc.__islocal) {
				return resolve(frm);
			}
			if (--tries <= 0) return resolve(null);
			setTimeout(tick, 100);
		};
		tick();
	});
}

// ⚠️ ONLY Sales Invoice and Purchase Invoice can start life as a return. `is_return`
// is READ-ONLY on Delivery Note and Purchase Receipt (measured on the meta), so a blank
// return of those cannot be created at all — theirs must be raised from the document
// being reversed, which is also what the return policy requires.
yht_custom.sales.RETURNABLE = ["Sales Invoice", "Purchase Invoice"];

// 🔴 A DELIVERY NOTE / PURCHASE RECEIPT RETURN CANNOT START AS A BLANK DOCUMENT.
// `is_return` is read_only = 1 AND no_copy = 1 on both, so nothing — not a route option,
// not a filtered list's "+ Add", not frappe.new_doc — can pre-tick it. Using the list's
// own Add button gives you an ordinary KSDN- note, silently, which is exactly what a
// branch user hit. The supported route is the mapper, which needs the source document,
// so ask for it.
//
// erpnext.stock.doctype.delivery_note.delivery_note.make_sales_return(source_name)
// erpnext.stock.doctype.purchase_receipt.purchase_receipt.make_purchase_return(source_name)
yht_custom.sales.STOCK_RETURN = {
	"Delivery Note": {
		method: "erpnext.stock.doctype.delivery_note.delivery_note.make_sales_return",
		title: __("Which delivery note are the goods coming back from?"),
		label: __("Delivery Note"),
	},
	"Purchase Receipt": {
		method: "erpnext.stock.doctype.purchase_receipt.purchase_receipt.make_purchase_return",
		title: __("Which receipt are the goods going back on?"),
		label: __("Purchase Receipt"),
	},
};

yht_custom.sales.new_stock_return = function (doctype) {
	const spec = yht_custom.sales.STOCK_RETURN[doctype];
	if (!spec) return;

	const d = new frappe.ui.Dialog({
		title: spec.title,
		fields: [
			{
				fieldname: "source",
				fieldtype: "Link",
				options: doctype,
				label: spec.label,
				reqd: 1,
				// Submitted, not itself a return, and not already fully returned — the
				// last one is what stops a second return against the same goods.
				get_query: () => ({
					filters: { docstatus: 1, is_return: 0, per_returned: ["<", 100] },
				}),
			},
		],
		primary_action_label: __("Create Return"),
		async primary_action(values) {
			d.hide();
			// open_mapped_doc calls the whitelisted mapper and routes to the result, so
			// the return arrives with is_return, return_against and the negative
			// quantities already set by erpnext itself.
			frappe.model.open_mapped_doc({ method: spec.method, source_name: values.source });

			// The picker still shows the FORWARD series until the document is saved —
			// the branch series is chosen server-side at before_insert, and it does
			// pick the right one (verified: KSDR-26-0025). But an operator seeing
			// KSDN- and getting KSDR- reads that as a bug, so show the answer now.
			const frm = await wait_for_new_form(doctype);
			if (!frm) return;
			try {
				const r = await frappe.call({
					method: "yht_custom.sales_flow.return_naming_series",
					args: { doctype: doctype },
				});
				const series = r && r.message;
				if (!series) return;
				const df = frm.get_field("naming_series");
				const options = (df && df.df.options ? df.df.options.split("\n") : []).map((o) => o.trim());
				if (options.includes(series)) await frm.set_value("naming_series", series);
			} catch (e) {
				// A bypass role or an unconfigured branch is not an error.
			}
		},
	});
	d.show();
};

yht_custom.sales.new_return = async function (doctype) {
	doctype = doctype || "Sales Invoice";
	if (!yht_custom.sales.RETURNABLE.includes(doctype)) {
		frappe.msgprint({
			title: __("Start From the Original"),
			indicator: "orange",
			message: __("A {0} return has to be raised from the document it reverses — open it and use <b>Create &gt; Return</b>.", [__(doctype)]),
		});
		return;
	}
	await frappe.new_doc(doctype);
	const frm = await wait_for_new_form(doctype);
	if (!frm) {
		frappe.show_alert({ message: __("Could not open a return — try again"), indicator: "red" });
		return;
	}
	await frm.set_value("is_return", 1);

	// The picker still shows the form's pre-filled invoice series: the real choice happens
	// server-side at before_insert. Ask for the answer and show it, or the operator sees
	// KSIN- on screen, saves, and gets KSSR- — which reads as a bug rather than a feature.
	let series = null;
	try {
		const r = await frappe.call({
			method: "yht_custom.sales_flow.return_naming_series",
			args: { doctype: doctype },
		});
		series = r && r.message;
	} catch (e) {
		// A bypass role or an unconfigured branch is not an error — leave the picker alone.
	}
	if (series) {
		const df = frm.get_field("naming_series");
		const options = (df && df.df.options ? df.df.options.split("\n") : []).map((o) => o.trim());
		if (options.includes(series)) await frm.set_value("naming_series", series);
	}

	frappe.show_alert({
		message: series
			? __("Credit note — this will be numbered {0}", [series])
			: __("Return ticked — this is a credit note. Pick the invoice it is against."),
		indicator: "blue",
	});
};

// The same entry point from the Sales Invoice list, so a workspace shortcut to the
// returns list lands one click away from creating one. A Workspace Shortcut is config
// only — it cannot run this — which is why the list carries the button instead.
//
// 🔴 DO NOT MERGE INTO `frappe.listview_settings["Sales Invoice"]`. erpnext's own
// `sales_invoice_list.js` opens with a WHOLESALE ASSIGNMENT to that key, and a doctype's
// list bundle is fetched when the list is first opened — i.e. AFTER `app_include_js`.
// Whatever we merge in at boot is discarded, silently: the button simply never appears,
// with no error. Attach from the router instead, which runs after the view is built no
// matter which file loaded first.
// The "New Return" button, on every list where a return can start.
//
// 🔴 DO NOT MERGE INTO `frappe.listview_settings[...]`. erpnext's own list JS assigns
// that key wholesale and a doctype's list bundle loads AFTER app_include_js, so anything
// merged in at boot is discarded silently. Attach from the router instead.
//
// 🔴 AND DO NOT ADD IT ONCE BEHIND A "done" FLAG. On a COLD load the list view clears its
// own inner toolbar AFTER the first render, so a button added before that is wiped —
// measured: the add ran, the flag was set, and the toolbar came back empty, while an
// in-app route to the same list worked. Key on the button's presence in the DOM and keep
// re-checking for a few seconds, which also covers the bundle still being in flight.
const RETURN_LISTS = {
	// A blank one CAN be a return: is_return is editable on the invoices.
	"Sales Invoice": () => yht_custom.sales.new_return("Sales Invoice"),
	"Purchase Invoice": () => yht_custom.sales.new_return("Purchase Invoice"),
	// A blank one CANNOT: is_return is read_only + no_copy on the stock documents, so
	// these ask for the source note and run erpnext's mapper.
	"Delivery Note": () => yht_custom.sales.new_stock_return("Delivery Note"),
	"Purchase Receipt": () => yht_custom.sales.new_stock_return("Purchase Receipt"),
};

frappe.router.on("change", () => {
	const route = frappe.get_route() || [];
	if (route[0] !== "List") return;
	const doctype = route[1];
	const action = RETURN_LISTS[doctype];
	if (!action) return;

	const label = __("New Return");
	let tries = 0;
	const attach = () => {
		const lv = window.cur_list;
		if (lv && lv.doctype === doctype && lv.page && lv.page.inner_toolbar) {
			const present = lv.page.inner_toolbar
				.find("button")
				.filter((i, b) => b.innerText.trim() === label).length;
			if (!present) {
				lv.page.add_inner_button(label, action);
			}
		}
		if (++tries < 20) setTimeout(attach, 300);
	};
	attach();
});
